packages/indexer-common/src/subgraphs.ts

```diff
@@ -323,3 +327,162 @@ export function isDeploymentWorthAllocatingTowards(
+    this.sleepDurationMillis = sleepDurationMillis
+    this.logger = logger
+    this.retries = retries
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public async checkedQuery<Data = any>(
+    subgraph: SubgraphQueryInterface,
+    query: DocumentNode,
+    variables?: Variables,
+  ): Promise<QueryResult<Data>> {
+    // Try to inject the latest block number into the original query.
+    let updatedQuery = query
+    try {
+      updatedQuery = mergeSelectionSets(query, blockNumberQuery)
+    } catch (err) {
+      const errorMsg = `Failed to append block number into ${this.subgraphName} query`
+      this.logger.error(errorMsg, { subgraph: this.subgraphName, query: print(query) })
+      throw new Error(errorMsg)
+    }
+
+    // Try obtaining a fresh subgraph query at most `this.retry` times
+    return this.checkedQueryRecursive<Data>(
+      updatedQuery,
+      subgraph,
+      this.retries,
+      variables,
+    )
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private async checkedQueryRecursive<Data = any>(
+    updatedQuery: DocumentNode,
+    subgraph: SubgraphQueryInterface,
+    retriesLeft: number,
+    variables?: Variables,
+  ): Promise<QueryResult<Data>> {
+    if (retriesLeft < 0) {
+      const errorMsg = `Max retries reached for ${this.subgraphName} freshness check`
+      this.logger.error(errorMsg, {
+        subgraph: this.subgraphName,
+        query: print(updatedQuery),
+      })
+      throw new Error(errorMsg)
+    }
+
+    // Obtain the latest network block number and subgraph query in parallel.
+    const subgraphQueryPromise = subgraph.query(updatedQuery, variables) as Promise<
+      QueryResult<Data> & BlockNumberInterface
+    >
+    const latestNetworkBlockPromise = this.provider.getBlockNumber()
+    const [subgraphQueryResult, latestNetworkBlock] = await Promise.all([
+      subgraphQueryPromise,
+      latestNetworkBlockPromise,
+    ])
+
+    // Return it early if query results contains errors
+    if (subgraphQueryResult.errors) {
+      return subgraphQueryResult
+    }
+
+    const latestIndexedBlock = subgraphQueryResult?.data?._meta?.block?.number
+    if (!latestIndexedBlock) {
+      const errorMsg = `Failed to infer block number for ${this.subgraphName} query`
+      this.logger.error(errorMsg, { query: print(updatedQuery) })
+      // Check for unexpected missing block data as a precaution.
+      throw new Error(errorMsg)
+    }
+
+    // Check subgraph freshness
+    const blockDistance = latestNetworkBlock - latestIndexedBlock
+    const logInfo = {
+      latestIndexedBlock,
+      latestNetworkBlock,
+      blockDistance,
+      freshnessThreshold: this.threshold,
+      subgraph: this.subgraphName,
+      retriesLeft,
+    }
+    this.logger.trace('Performing subgraph freshness check', logInfo)
+
+    if (blockDistance < 0) {
+      // Invariant violated: Subgraph can't be ahead of network latest block
```

Comment:
If the subgraph is ahead of the provider I think that's ok. SUPER FRESH!
Let's just warn here?